doc/qbk/21_connection_pool.qbk

```diff
@@ -0,0 +1,251 @@
+]
+
+This is how you configure pool sizes:
+
+[connection_pool_configure_size]
+
+
+
+[heading Session state]
+
+MySQL connections hold state. You change session state when you prepare statements,
+create temporary tables, start transactions, or set session variables. When using
+pooled connections, session state can be problematic: if not reset properly,
+state from a previous operation may affect subsequent ones.
+
+After you return a connection to the pool, it uses [refmem any_connection async_reset_connection]
+to wipe session state before the connection can be obtained again. This will deallocate
+prepared statements, rollback uncommited transactions and clear variables.
+In particular, you don't need to call [refmem any_connection async_close_statement]
+to deallocate statements.
+
+Resetting a connection is cheap but entails a cost (a roundtrip to the server).
+If you've used a connection and you know that you didn't mutate session state,
+you can use [refmem pooled_connection return_without_reset] to skip resetting.
+For instance:
+
+[connection_pool_return_without_reset]
+
+Connection reset happens in the background, after the connection has been
+returned, so it does not affect latency. If you're not sure if an operation
+affects state or not, assume it does.
+
+
+[heading Character sets]
+
+[warning
+    The current implementation, together with unintuitive MySQL defaults,
+    can yield to surprising behavior.
+]
+
+When using pooled connections, [*all connections use the server's default character set].
+This is because [refmem any_connection async_reset_connection] discards the character set
+options specified during connection establishment.
+
+You can obtain this character set by running:
+
+[!teletype]
+```
+SELECT @@global.character_set_client, @@global.character_set_connection, @@global.character_set_results;
+```
+
+MySQL v8.0+ defaults to `utf8mb4`, but older MySQL and MariaDB servers default to
+`latin1`, which is not usually what you want. Issue a __SET_NAMES__ statement
+after you get a connection to change it. [@https://github.com/boostorg/mysql/issues/187 This issue]
+tracks solving this limitation.
+
+
+[heading Connection lifecycle]
+
+The behavior already explained can be summarized using a state model like the following:
+
+[$mysql/images/pooled_connection_lifecycle.svg [align center]]
+
+In short:
+
+* When a connection is created, it goes into the `pending_connect` state.
+* Connection establishment is attempted. If it succeeds, the connection becomes `iddle`.
+  Otherwise, it stays `pending_connect`, and another attempt will be performed
+  after [refmem pool_params retry_interval] has ellapsed.
+* `iddle` connections can be retrieved by [refmem connection_pool async_get_connection],
+  and they become `in_use`.
+* If a connection is returned by [refmem pooled_connection return_without_reset],
+  it becomes `iddle` again.
+* If a connection is returned by [reflink pooled_connection]'s destructor, it becomes
```

Comment:
Because you may want to steal a connection (upgrade it to a websocket) and doing implicit things is dtors just seems off. The main difference to mysql however is that you'll use the pooled connection through an algorithm like `http::async_get` about 97% of the time, so users don't need to explicitly return the thing either.